### PR TITLE
[USDU-407] Add automated tests for Analytics 

### PR DESCRIPTION
**Pull request by lee-aandrew**
## Purpose of this PR
Analytics have been added for our USD package giving us various information about the usage of the package
We thought it would have be needed to have some automated test cases for the analytics

**Ticket/Jira #:**
[USDU-407]

## Testing

**Functional Testing status:**
Adds functional testcases for the analytics for the following parts:
- Import
- Export
- Re-import (refresh / update)

Could not add planned Recorder analytics automated testing as a lot of recorder aspect is private / internal, and unless I use reflections I cannot instantiate a lot of the classes that I would need to set up the test
LMK if you guys think its needed

**Performance Testing status:**

## Overall Product Risks

**Complexity:**

**Halo Effect:**

## Additional information

**Note to reviewers:**

**Reminder:**

- [ ] Add entry in CHANGELOG.md _(if applicable)_

### CODE REVIEW

**Comment by SimonBoorer** [comment]
Are you aware the tests are failing on 2023.1 and newer?
> Library\PackageCache\com.unity.formats.usd.tests@090deec1ff8c\Tests\Editor\EditorAnalyticsBaseFixture.cs(8,7): error CS0246: The type or namespace name 'Newtonsoft' could not be found (are you missing a using directive or an assembly reference?)

**Review by SimonBoorer — APPROVED** [review]

File: package/com.unity.formats.usd/Tests/Common/BaseFixture.cs [hunk 1]
```diff
@@ -75,6 +75,8 @@ public string GetPrefabPath(string prefabName = null, bool resource = false)
                 prefabName += ".prefab";
             }
 
+            var test = Path.Combine(ArtifactsDirectoryRelativePath, resource ? "Resources" : "", prefabName);
```
Comment (michaeljblain) [inline package/com.unity.formats.usd/Tests/Common/BaseFixture.cs]: Are these 2 lines just for your personal debugging and can they be removed?

File: package/com.unity.formats.usd/Tests/Common/BaseFixture.cs [hunk 2]
```diff
@@ -126,7 +137,14 @@ public void CleanupTestArtifacts()
...
             if (Directory.Exists(ArtifactsDirectoryFullPath))
             {
-                Directory.Delete(ArtifactsDirectoryFullPath, true);
+                try
+                {
+                    Directory.Delete(ArtifactsDirectoryFullPath, true);
+                }
+                catch
+                {
+                    Debug.Log("Artifact Clean up has failed - This should not happen in most cases, but even if so, the test case should not be affected.");
```
Comment (michaeljblain) [inline package/com.unity.formats.usd/Tests/Common/BaseFixture.cs:59]: Can we feed back further detail on the failure by also logging the exception?

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
+# if UNITY_2023_2_OR_NEWER
```
Comment (michaeljblain) [inline package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: Are we restricted to >= 2023.2 due to Vicky's Analytics fix on trunk? Or is there some other reason? Should we consider adding a note here about why it is 2023.2 or newer?

**Review by michaeljblain — CHANGES_REQUESTED** [review]
Some suggestions. Looks like tests are still failing on trunk.

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs (hunk 3 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: yes the restriction is due to her fix only being on 2023.2+

File: package/com.unity.formats.usd/Tests/Common/BaseFixture.cs (hunk 1 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Common/BaseFixture.cs]: yes i forgot i left them in, sorry! Deleted!

File: TestProject/Usd-Development/Packages/manifest.json [hunk 4]
```diff
@@ -1,11 +1,12 @@
 {
   "dependencies": {
     "com.unity.analytics": "3.5.3",
+    "com.unity.editor-analytics-debugger": "0.1.0",
```
Comment (vickycl) [inline TestProject/Usd-Development/Packages/manifest.json]: Is this ok in 2019.4?

**Comment by lee-aandrew** [comment]
~Stuck due to problem on the analytics side - Mike Stowell is looking into the issue~

Requires Mike's fixes: [Link](https://github.cds.internal.unity3d.com/unity/unity/pull/33470)

File: .yamato/package-test.yml [hunk 5]
```diff
@@ -13,7 +13,11 @@ test_{{ platform.name }}_{{ editor.version }}:
     UPMCI_PKG: "upm-ci-utils@stable"
   commands:
       - npm install {% if platform.name == "win" %}"%UPMCI_PKG%"{% else %}"$UPMCI_PKG"{% endif %} -g --registry {{ registry.npm }}
-      - upm-ci package test -u {{ editor.version }} --package-path package/com.unity.formats.usd --type package-tests --extra-utr-arg="--api-profile=NET_4_6"
+      {% if editor.version == "trunk" %}
+        - upm-ci package test -u {{ editor.version }} --package-path package/com.unity.formats.usd --type package-tests --extra-utr-arg="--api-profile=NET_4_6" --extra-editor-arg=-editorAnalyticsTestMode
```
Comment (lee-aandrew) [inline .yamato/package-test.yml]: This is behind an if statement as there was a bug in the analytics side where if this was set for versions less than trunk, it would crash the editor

Comment (vickycl) [reply .yamato/package-test.yml]: I don't think we should specify 'trunk' here as it isn't robust for the future. Eventually trunk will be newer that 2023 and the '23 LTS won't be tested/ the tests will run without this editor arg being set. I'm not sure what the correct answer is though :)

Comment (vickycl) [inline .yamato/package-test.yml]: does the `--extra-editor-arg=-editorAnalyticsTestMode` alleviate the need for UNITY_THISISABUILDMACHINE? ie does it set the runmode to test?

File: package/com.unity.formats.usd/Tests/Common/Data/Invalid/Kiki_Blendshape.usd
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Common/Data/Invalid/Kiki_Blendshape.usd:1]: I can't see what is included here, but it's quite a big asset to include for the sake of a test, and I'm also not sure whether it can be shared publicly in this way?

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        [SetUp]
+        public void ResetEventList()
+        {
+            DebuggerEventListHandler.ClearEventList();
+            InitUsd.Initialize();
+        }
+
+        public IEnumerator WaitForUsdAnalytics<T>(string expectedType, System.Action<AnalyticsEvent> expectedEvent, float attemptFrameCountLimit = 300) where T: UsdAnalyticsEvent
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: This frame count limit is very high- are you sure it needs to be this high?

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+            {
+                AnalyticsEvent expectedAnalyticsEvent = new AnalyticsEvent();
+                foreach (var concattedEvent in DebuggerEventListHandler.fetchEventList())
+                {
+                    if (!concattedEvent.Contains(expectedType))
+                    {
+                        continue;
+                    }
+
+                    Debug.Log($"Raw Log: {concattedEvent}");
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: Please remove the logging in every test run- it adds a lot of noise and some overhead. If the logging is necessary, maybe just do it on test failure?

File: package/com.unity.formats.usd/Tests/Editor/EditorImportAnalyticsTests.cs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            ImportAs(m_scene, importMethod);
+
+            AnalyticsEvent expectedEvent = null;
+            yield return (WaitForUsdAnalytics<UsdAnalyticsEventImport>(UsdAnalyticsTypes.Import, (AnalyticsEvent waitedEvent) => {
+                expectedEvent = waitedEvent;
+            }));
+
+            Assert.IsNotNull(expectedEvent);
+
+            var expectedUsdEventMsg = ((UsdAnalyticsEventImport)expectedEvent.usd).msg;
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorImportAnalyticsTests.cs]: Nitpick: This isn't the 'expected', it's the 'actual', right?

File: package/com.unity.formats.usd/Tests/Editor/EditorReImportAnalyticsTests.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            yield return (WaitForUsdAnalytics<UsdAnalyticsEventReImport>(UsdAnalyticsTypes.ReImport, (AnalyticsEvent waitedEvent) => {
+                reImportEvent = waitedEvent;
+            }));
+
+            Assert.IsNotNull(reImportEvent);
+            var reImportUsdEventMsg = ((UsdAnalyticsEventReImport)reImportEvent.usd).msg;
+
+            Assert.IsTrue(reImportUsdEventMsg.Succeeded, "Expected True for re-import Successful");
+            Assert.Greater(reImportUsdEventMsg.TimeTakenMs, 0, "Expected Time Taken MS should be greater than 0 for Re-Import");
+            Assert.AreEqual(importUsdEventMsg.IncludesSkel, reImportUsdEventMsg.IncludesSkel);
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorReImportAnalyticsTests.cs:143]: I think it would be better to check the Reimport result against a constant expected value instead of relying on the Import message, because if something went wrong that broke both Import and Reimport, comparing the two against each other wouldn't show anything wrong

File: package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs [hunk 10]
```diff
@@ -40,7 +42,11 @@ public static class UsdEditorAnalytics
         // Universal USD Analytics
         const string k_UsageEventName = "USDPackageUsage";
 
-        struct UsageAnalyticsData
+        public interface IUsdAnalyticsData
```
Comment (vickycl) [inline package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs]: This change seems to be unnecessary now since you implemented the UsdAnalyticsEvent class in the tests?

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        }
+    }
+
+    public class AnalyticsEvent
+    {
+        public SharedAnalyticsEvent shared;
+        public UsdAnalyticsEvent usd;
+    }
+
+    public class SharedAnalyticsEvent
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: The shared event is no longer used anywhere- can it be removed now?

File: package/com.unity.formats.usd/Tests/Editor/EditorExportAnalyticsTests.cs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+            AnalyticsEvent exportEvent = null;
+            yield return (WaitForUsdAnalytics<UsdAnalyticsEventExport>(UsdAnalyticsTypes.Export, (AnalyticsEvent waitedEvent) => {
+                exportEvent = waitedEvent;
+            }));
+
+            Assert.IsNotNull(exportEvent);
+            var exportUsdEventMsg = ((UsdAnalyticsEventExport)exportEvent.usd).msg;
+
+            Assert.IsTrue(exportUsdEventMsg.Succeeded, "Expected True for re-import Successful");
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorExportAnalyticsTests.cs]: think this should be 'export successful'

File: package/com.unity.formats.usd/Tests/Editor/EditorExportAnalyticsTests.cs [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorExportAnalyticsTests.cs]: As above :)

File: package/com.unity.formats.usd/Tests/Editor/EditorExportAnalyticsTests.cs [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+            AnalyticsEvent exportEvent = null;
+            yield return (WaitForUsdAnalytics<UsdAnalyticsEventExport>(UsdAnalyticsTypes.Export, (AnalyticsEvent waitedEvent) => {
+                exportEvent = waitedEvent;
+            }));
+
+            Assert.IsNotNull(exportEvent);
+            var exportUsdEventMsg = ((UsdAnalyticsEventExport)exportEvent.usd).msg;
+
+            Assert.IsFalse(exportUsdEventMsg.Succeeded, "Expected Fail for re-import Successful");
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorExportAnalyticsTests.cs]: As above :)

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs (hunk 11 above)
Comment (vickycl) [reply package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: In which case can we remove a layer of abstraction and use the UsdAnalyticsEvent directly instead of the AnalyticsEvent wrapper? To simplify lines like `var exportUsdEventMsg = ((UsdAnalyticsEventExport)exportEvent.usd).msg;` a bit, as they're hard to parse

Comment (vickycl) [reply package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: for eg `var exportUsdEventMsg = ((UsdAnalyticsEventExport)exportEvent).eventData;` in that case would be easier to reason at a glance what it's getting

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    {
+        // check if the class for this is in Unity source (if its a bit convoluted just make it)
+        object common;
+    }
+
+    public class UsdAnalyticsEvent
+    {
+        public string type;
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs:114]: Am I understanding correctly that this actually stores the event name string eg `editor.USDFileImport`, or is it actually the C# type name of the event? If it is the former I'd suggest a more explicit variable name :)

**Review by vickycl — COMMENTED** [review]
Small fixes and clean up suggested, but otherwise good so far!

File: TestProject/Usd-Development/Packages/manifest.json (hunk 4 above)
Comment (vickycl) [reply TestProject/Usd-Development/Packages/manifest.json]: Oops, this is leftover from when I initially started the review 😂

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+    public class AnalyticsEvent
+    {
+        public SharedAnalyticsEvent shared;
+        public UsdAnalyticsEvent usd;
+    }
+
+    public class SharedAnalyticsEvent
+    {
+        // check if the class for this is in Unity source (if its a bit convoluted just make it)
```
Comment (vickycl) [inline package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: What does this comment mean?

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs (hunk 15 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs:114]: nono its the former, its the event name, i'll change it to something like "name" or "eventName"

File: .yamato/package-test.yml (hunk 5 above)
Comment (lee-aandrew) [reply .yamato/package-test.yml]: I realized i set the "UNITY_THISISABUILDMACHINE" but it never got passed over to the yamato call, and i believe our test machines already have the "UNITY_THISISABUILDMACHINE" set to true

File: package/com.unity.formats.usd/Tests/Common/Data/Invalid/Kiki_Blendshape.usd
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Common/Data/Invalid/Kiki_Blendshape.usd:1]: Thats a good point, i just needed an invalid USD file for a failure, in this case blendshapes, i'll look for another asset

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs (hunk 16 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: It can be removed, but i meant to check if theres a class with the same structure as "SharedAnalyticsEvent" already defined within Unity, and if so, use that instead of creating it here

File: package/com.unity.formats.usd/Tests/Editor/EditorImportAnalyticsTests.cs (hunk 8 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Editor/EditorImportAnalyticsTests.cs]: good point - will change the names

File: .yamato/package-test.yml (hunk 5 above)
Comment (lee-aandrew) [reply .yamato/package-test.yml]: Yeah I think that will have to be changed in the future, once neutron stuff becomes the new trunk (iirc 2023 is the last "classic" unity before neutron)

If there is a 2023.4 of "classic" Unity that becomes the new trunk, in that case we will have to add 2023.3 as part of this if statement

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs (hunk 15 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs:114]: actually i cant change this as the result json's variable is called "type" and as far as i know Unity's JsonUtility does not have something similar to `JsonProperty`

File: package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs (hunk 10 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Common/UsdEditorAnalytics.cs]: Good point, will revert

File: package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs (hunk 6 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Editor/EditorAnalyticsBaseFixture.cs]: Changed it to use c#'s stopwatch  for shorter + more accurate measurement

File: package/com.unity.formats.usd/Tests/Editor/EditorReImportAnalyticsTests.cs (hunk 9 above)
Comment (lee-aandrew) [reply package/com.unity.formats.usd/Tests/Editor/EditorReImportAnalyticsTests.cs:143]: The point of this specific test case is that for the initial import and re-import the properties are equal
If theres a problem with the initial import, it would have been caught with the import tests, and if theres a problem with the re-import it would have been caught in the re-import specific tests
This testcase is to make sure that on re-import, some properties do not change for simple files like the test file
